oomd: allow ManagedOOMPreference on all cgroups

enr0n opened the pull request:
Firstly, I consider this a sort of RFC because I know that it cannot be merged in the current state. But, I think making this change is worthwhile, specifically for desktop environments using systemd-oomd.

The motivation behind these patches is to allow desktop environments to have some control over which applications are targeted (or avoided) by systemd-oomd when it needs to act on candidate cgroups. In the current state, the `ManagedOOMPreference` property is ignored for any cgroups which are not owned by root. This is unfortunate for desktop environments where applications are launched as descendants of `user@.service`.

In practice, when systemd-oomd needs to act on desktop systems, user applications such as browsers are killed first. For desktop users, it is very likely that their browser (or some other desktop application) is their most critical program, and it may make sense to de-prioritize the browser's cgroup as a kill candidate. For this reason, I think it is important that systemd-oomd honors `ManagedOOMPreference` on cgroups which are not owned by root.

Based on the conversation in https://github.com/systemd/systemd/pull/18401/, I understand why this restriction on `ManagedOOMPreference` exists -- to "prevent situations where an unprivileged user can set their own cgroups lower in the kill priority than another user's". The "RFC" part of this PR is then, how can we retain the ability of root to have discretion over `ManagedOOMPreference` without completely restricting the property from cgroups owned by non-root users?

anitazha commented:
We also had https://github.com/systemd/systemd/pull/19007 that is a bit outdated now since it was written before we had systemd-oomd settings for user services. I need to recall the idea I had before but it was something like a permission/UID check between the cgroup with xattrs and the user manager that "owns" the systemd-oomd settings. That way we can ignore the xattrs if we see it on e.g. the system.slice level, but respect it if it's at the user@.slice level. (I'll be traveling for the next week or so so my responses will be spotty.)

enr0n commented:
> We also had #19007 that is a bit outdated now since it was written before we had systemd-oomd settings for user services. I need to recall the idea I had before but it was something like a permission/UID check between the cgroup with xattrs and the user manager that "owns" the systemd-oomd settings. 

Similarly to the way this is done for `ManagedOOMSwap` and `ManagedOOMMemoryPressure` in `process_managed_oom_message`?

> That way we can ignore the xattrs if we see it on e.g. the system.slice level, but respect it if it's at the [user@.slice](mailto:user@.slice) level. (I'll be traveling for the next week or so so my responses will be spotty.)

I don't think I follow, can you please elaborate? When exactly would we ignore the xattrs?

enr0n commented:
> Similar to the ManagedOOM*= settings themselves, the xattrs should also be grouped according to the requester. Example: if we're processing candidates for a cgroup owned by root (e.g. pid1, service manager), we might see `user.oomd_omit` xattr set on `/user.slice/user-1000.slice/user@1000.service/app.slice/myuserservice.service`. In this case we would ignore that xattr because it was set by the user manager for UID 1000. This would avoid unprivileged users affecting the kill priority of an ancestor cgroup.

Okay, I think I understand. Does that mean, for example, that if system was configured with `ManagedOOM{Swap,MemoryPressure}=kill` on `user@.service`, but not on any ancestor cgroups, then unprivileged users would be able to configure `ManagedOOMPreference`? And, if a system was configured with `ManagedOOM{Swap,MemoryPressure}=kill` on e.g. `-.slice`, then unprivileged users would **never** be able to configure `ManagedOOMPreference`?

enr0n commented:
I pushed new commits to implement what we have discussed, based on my understanding of the approach.

enr0n commented:
I see the failing tests, and I will push changes to address that.

enr0n commented:
The remaining test failure appears to be unrelated to these changes.

anitazha commented on src/oom/oomd-util.c:
```diff
@@ -355,10 +357,10 @@ int oomd_cgroup_context_acquire(const char *path, OomdCGroupContext **ret) {
-        else if (uid == 0) {
+        if ((r = cg_get_owner(SYSTEMD_CGROUP_CONTROLLER, path, &cg_uid)) < 0 ||
+            (r = cg_get_owner(SYSTEMD_CGROUP_CONTROLLER, monitored_ancestor_path, &ancestor_cg_uid)) < 0)
+                log_debug_errno(r, "Failed to get owner/group from %s and %s: %m", path, monitored_ancestor_path);
```
and -> and/or (?)
Alternatively split these conditions up and log the messages separately

enr0n commented on man/systemd.resource-control.xml:
```diff
@@ -1150,9 +1150,16 @@ DeviceAllow=/dev/loop-control
+          the unit's cgroup owner, and the owner of the monitored ancestor cgroup are the
+          same. For example, if <filename>-.slice</filename> is monitored by <command>systemd-oomd</command>,
+          then these extended attributes would be ignored if set on descendants of
+          <filename>/user.slice/user-1000.slice/user@1000.service/</filename>. However, if no ancestors of
+          <filename>/user.slice/user-1000.slice/user@1000.service/</filename> are monitored by
```
> > However, if no ancestors of /user.slice/user-1000.slice/user@1000.service/ are monitored by...
> 
> This should be allowed as long as the avoid/omit cgroup is a candidate for a monitored cgroup that has the same cgroup owner. So for example, if we have something like this:
> 
> * -.slice # monitoring for pressure threshold 60%
>   
>   * user.slice
>     
>     * user-1000.slice
>       
>       * [user@1000.service](mailto:user@1000.service) # monitoring for pressure threshold 30%
>         
>         * app.slice
>           
>           * my.service # oomd_omit
> 
> xattrs on `my.service` can be respected if it is a candidate for `user@1000.service`. But if we're calculating the candidates for `-.slice`, then oomd xattrs on `my.service` should be ignored. Basically regardless of whether `user@1000.service` has an ancestor being monitored or not. Does that make sense?

Yes, this is a good clarifying example. It is more clear looking at the memory pressure scenario because `oomd_pressure_above` does initial filtering of the cgroups which means (using the example above) `user@1000.service` could be a target, but not `-.slice`. But, with swap kill every monitored cgroup is an initial candidate and will be sorted in `oomd_kill_by_swap_usage`, which led me to the monitored ancestor logic.

> This last part should be reworded to reflect that (maybe remove the stuff about "no ancestors"). I think this might also simplify some of the code such that the check between UIDs would only need to happen during candidate selection/ordering.

Yes, I think this should simplify the code quite a bit. Thanks for taking a look.

enr0n commented:
I believe the remaining test failures are unrelated (I know the focal tests are infrastructure related, not sure about CentOS CI). PTAL.

anitazha reviewed (commented):
Sorry for the delay. I did one round over the PR and it looks great. I would want to do another round tomorrow to be sure.
However I'm not sure we want the swap kill candidate change (commented inline).

enr0n commented on src/oom/oomd-manager.c:
```diff
@@ -404,21 +406,35 @@ static int monitor_swap_contexts_handler(sd_event_source *s, uint64_t usec, void
-                        log_notice_errno(r, "Failed to kill any cgroup(s) based on swap: %m");
+
+                n = oomd_sort_cgroup_contexts(m->monitored_swap_cgroup_contexts,
+                                              compare_cgroup_path_and_swap_usage,
```
Yeah, I see what you mean. How exactly should we determine when a swap kill candidate's OOM preference should be honored then? Since swap kill candidates are sorted all at once, regardless of which monitored cgroup context they are descendant of, we cannot use the same logic as the memory pressure scenario (this commit was an attempt to make that possible).

If we just dropped "oomd: calculate swap kill candidates one monitored cgroup at a time" from this PR, then the existing behavior (only cgroups owned by root can set `ManagedOOMPreference`) would remain for the swap kill scenario.

anitazha replied:
> If we just dropped "oomd: calculate swap kill candidates one monitored cgroup at a time" from this PR, then the existing behavior (only cgroups owned by root can set ManagedOOMPreference) would remain for the swap kill scenario.

I think this makes the most sense. Given that swap kill thresholds are globally controlled, root should be able to use `ManagedOOMPreference=`. Non-root cgroups that don't want to get killed could have someone privileged set up a more fine grained policy where certain cgroups are omitted altogether from the candidate list. For example:
```
/ [ManagedOOMSwap=auto]
  /system.slice [ManagedOOMSwap=kill]
  /user.slice [ManagedOOMSwap=auto]
    /user.slice/user-1000.slice [ManagedOOMSwap=kill]
    /user.slice/user-3000.slice [ManagedOOMSwap=auto]
```
Which would skip `/user.slice/user-3000.slice` as a swap kill candidate, but still include everything else

enr0n replied:
Okay, I have removed this commit from the PR. I believe the test failures are unrelated.

anitazha commented on src/oom/test-oomd-util.c:
```diff
@@ -451,6 +406,90 @@ static void test_oomd_sort_cgroups(void) {
+                assert_se(oomd_get_cgroup_oom_preference(ctx, NULL) == 0);
+                assert_se(cg_set_xattr(SYSTEMD_CGROUP_CONTROLLER, cgroup, "user.oomd_omit", "1", 1, 0) >= 0);
+                assert_se(cg_set_xattr(SYSTEMD_CGROUP_CONTROLLER, cgroup, "user.oomd_avoid", "1", 1, 0) >= 0);
+        }
+
+        /* omit takes precedence over avoid when both are set to true */
+        if (test_xattrs) {
```
nit: can combine these `if` statements

enr0n replied:
Thanks, fixed!

anitazha reviewed (approved):
lgtm. minor nit

poettering commented on src/oom/oomd-util.c:
```diff
@@ -142,10 +143,49 @@ bool oomd_swap_free_below(const OomdSystemContext *ctx, int threshold_permyriad)
+                ctx->preference = r == 1 ? MANAGED_OOM_PREFERENCE_AVOID : ctx->preference;
+
+                r = cg_get_xattr_bool(SYSTEMD_CGROUP_CONTROLLER, ctx->path, "user.oomd_omit");
+                if (r == -ENOMEM)
```
ditto

poettering commented on src/oom/oomd-util.c:
```diff
@@ -142,10 +143,49 @@ bool oomd_swap_free_below(const OomdSystemContext *ctx, int threshold_permyriad)
+                r = cg_get_xattr_bool(SYSTEMD_CGROUP_CONTROLLER, ctx->path, "user.oomd_avoid");
+                if (r == -ENOMEM)
+                        return r;
+                ctx->preference = r == 1 ? MANAGED_OOM_PREFERENCE_AVOID : ctx->preference;
```
doesn't matter, but I think it would be nicer to do this check as `r > 0` rather than `r == 1`.

When doing "ternary" booleans, i.e. booleans that are true or false, or indicate some error, we say > 0 means true, == 0 means false and < 0 means error. Hence to be systematic here it would be good to do write the check above as > 0, even if of course 1 is the only relevant value here.

poettering commented:
(my comments dont matter much, I am aware the points I raised are already existent in the code before this commit — and mostly bikeshedding – but it would still be good to get this fixed)

enr0n commented:
> (my comments dont matter much, I am aware the points I raised are already existent in the code before this commit — and mostly bikeshedding – but it would still be good to get this fixed)

Thanks for your review. I believe I addressed all of your comments in the last push.

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -142,10 +143,53 @@ bool oomd_swap_free_below(const OomdSystemContext *ctx, int threshold_permyriad)
+                if (r == -ENOMEM)
+                        return r;
+                if (r < 0 && r != -ENODATA)
+                        log_debug_errno(r, "Failed to get xattr user.oomd_avoid: %m");
```
Please add `", ignoring"` to the log message for ignored error.

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -142,10 +143,53 @@ bool oomd_swap_free_below(const OomdSystemContext *ctx, int threshold_permyriad)
+                 * as an optional feature of systemd-oomd (and the system might not even support them). */
+                r = cg_get_xattr_bool(SYSTEMD_CGROUP_CONTROLLER, ctx->path, "user.oomd_avoid");
+                if (r == -ENOMEM)
+                        return r;
```
Please use `log_oom_debug()`.

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -142,10 +143,53 @@ bool oomd_swap_free_below(const OomdSystemContext *ctx, int threshold_permyriad)
+
+                r = cg_get_xattr_bool(SYSTEMD_CGROUP_CONTROLLER, ctx->path, "user.oomd_omit");
+                if (r == -ENOMEM)
+                        return r;
+                if (r < 0 && r != -ENODATA)
+                        log_debug_errno(r, "Failed to get xattr user.oomd_omit: %m");
```
Same here. Please use `log_oom_debug()` and append `", ignoring"`.

enr0n replied:
Done, thanks!

yuwata reviewed (approved):
LGTM. But several minor comments.

enr0n commented:
I think the remaining failure (focal amd64 autopkgtest) is unrelated.

emansom commented:
I usually have a ridicilious number of tabs both in Chromium and Firefox open, and would prefer `systemd-oomd` seeing that memory hog as a candidate to kill.

Current behavior yields a livelock eventually.

How would one go about configuring that?
